working/macros/feature-specification.md

```diff
@@ -550,6 +550,116 @@ This does have two interesting and possibly unexpected consequences:
 
 **TODO**: Define the API for adding metadata to existing declarations (#1931).
 
+## Aspect Macros
+
+**NOTE**: This feature is a work in progress, and has many TODOs.
+
+An AspectMacro is used by another macro, to generate shared helper code which
+will not be duplicated across every library where that helper was needed.
+
+Concretely, let's consider the following example:
+
+```dart
+@fromJsonExtension
+class A {}
+
+@fromJsonExtension
+class B {
+  final A a;
+
+  B(this.a);
+}
+
+// Generated extension from the macro on A
+extension AFromJson on Map<String, Object?> {
+  A toA();
+}
+
+// Generated extension from the macro on B
+extension BFromJson on Map<String, Object?> {
+  A toA(); // This also needs to generate a helper for converting `A`!
+
+  B toB();
+}
+```
+
+Both of these macro applications need to generate the same shared code for
+converting a Map into an instance of `A` (since `B` has a field of that type),
+but they can't actually _see_ what each other has generated.
+
+In the worst case, this would result in them both outputting conflicting
+members, and in the best case they would end up generating duplicate code,
+causing unnecessary code bloat.
+
+Aspect Macros exist to solve this problem, by allowing macros to share these
+implementations. This is done by associating a single, canonical declaration
+with an Aspect Macro Instance + Declaration pair, and giving back an opaque
+identifier for the resulting declaration.
```

Comment:
What if there is no logical declaration to attach the aspect macro to? Imagine you have a macro that needs some big blob of generated static data. You only want one copy of that data in the program, but it's not specific to any particular declaration, it's just shared by all applications of the macro.

Can the declaration be optional?

Reply:
This feels related to your other comment, in this case it is likely generating stuff based on some sort of input data, possibly produced by other macros, which is then likely aggregated together in some fashion?

Reply:
Even without aggregating, I can imagine that a macro might want to generate some helper function whose body maybe depends on some data known by the macro. but that doesn't really come from a particular *declaration*.

For example, imagine a `fieldHash` macro that generates a `hashCode` member on a class. It does so in terms of a series of helper functions like `hash1(Object? a)`, `hash2(Object? a, Object? b)`, etc. A `hash_n_()` helper with some given arity is only needed if the macro is applied to a class with that many fields. The macro author could hand author a helper library that contains `hash_n_()` methods up to some fixed arity and rely on the tree-shaker to discard all the unused ones. But they're essentially copy/paste of each other.

Instead, it might be nice if the `fieldHash` macro could just generate those helper functions on demand for the specific arities it needs. We still definitely want these functions to be deduplicated. But they aren't applied to any particular *declaration*.

Another way to think about it is that aspect macros here are essentially functions that take in some parameters and produce a declaration. They are memoized in that if you call the same aspect macro with the same parameters, the redundant declaration gets discarded. In the design here, an aspect macro can only be parameterized on one thing: a declaration. And it *must* be parameterized on that.

My suggestion is to generalize that to allow them to be parameterized on basically any set of data (as long as we can reliably send it over an isolate and test for equality, etc.). In the above (very contrived) example, it would just be parameterized on an int representing the arity.

Reply:
Ah yeah the hash example is interesting, and I could see other similar uses as well